### ISSUE TITLE
Tests connection.switch.port_link and connection.switch.port_speed, make sure they works with Cisco switch

### ISSUE THREAD

**Issue by spectre-storm** [faucetsdn/daq#233]

**Comment by spectre-storm** [comment]
connection.switch.port_link
connection.switch.port_speed

**Comment by pisuke** [comment]
@jhughesbiot I've configured the switch module for both the AT x230 and the Cisco 9300 switch, but so far I haven't been able to make them work.

The attached files are the activate.log outputs from the test communication with the switch. 

For both the AT switch there seem to be some problems with the communication timing out.
See these activate.log outputs.

[at_x230_activate.log](https://github.com/faucetsdn/daq/files/4470059/at_x230_activate.log)

[cisco_9300_activate.log](https://github.com/faucetsdn/daq/files/4470061/cisco_9300_activate.log)

Here are my system.conf files for the AT x230
```
# DAQ configuration for the Allied Telesis x230 switch
source misc/system.conf
site_description="Allied Telesis x230"
ext_dpid=0x42
ext_ctrl=enxb49cdff312cf
ext_intf=enx645aedf40fb3
ext_ofpt=6653
ext_ofip=192.168.1.10/24
ext_addr=192.168.1.1
ext_loip=192.168.1.@/24
sec_port=15
host_tests=misc/all_tests.conf
switch_model=ALLIED_TELESIS_X230
```

and for the Cisco 9300
```
# DAQ configuration for the Allied Telesis x230 switch
source misc/system.conf
site_description="Cisco 9300"
ext_dpid=0xABCDEF1234
ext_ctrl=enx645aedf40df6
ext_intf=enxb49cdff3015b
ext_ofpt=6653
ext_ofip=192.168.10.11/24
ext_addr=192.168.10.10
ext_loip=192.168.10.@/24
sec_port=24
host_tests=misc/all_tests.conf
switch_model=CISCO_9300
```

Here is the output of a test run, showing the time out error.
```
francesco@arup-daq-lab:~/Code/daq/pisuke/daq$ cmd/run -s
Activating venv
Loading config from local/system.conf into inst/config/system.conf
Starting Mon 13 Apr 16:01:33 BST 2020
Clearing previous state...
No LSB modules are available.
Running as root...
No LSB modules are available.
Activating venv
Loading config from local/system.conf into inst/config/system.conf
HEAD is now at f13396aa Merge pull request #3476 from mab68/orderedacl1
HEAD is now at b03e4666 Built docker images (#57)
Release version 1.1.0-26-g40abe0d
ovsdb-server is running with pid 423
ovs-vswitchd is running with pid 525
DAQ autostart bin/switch_local
Activating venv
Loading config from local/system.conf into inst/config/system.conf
Pulled cplane_mac 64:5a:ed:f4:0d:f6 from enx645aedf40df6
Cleaning old setup...
Creating ovs-link interfaces...
Creating local-link interfaces...
Creating local bridge...
Configuring ctrl-swy with 192.168.10.11/24
Bridging enx645aedf40df6 to ctrl-br
Checking external connection to 192.168.10.10
PING 192.168.10.10 (192.168.10.10) 56(84) bytes of data.
64 bytes from 192.168.10.10: icmp_seq=1 ttl=255 time=0.428 ms
64 bytes from 192.168.10.10: icmp_seq=2 ttl=255 time=0.438 ms

--- 192.168.10.10 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1000ms
rtt min/avg/max/mdev = 0.428/0.433/0.438/0.005 ms
DAQ autoclean ip link del ctrl-pri
DAQ autoclean ip link del ctrl-swx
DAQ autoclean ovs-vsctl --if-exists del-br ctrl-br
Done with local switch setup.
Entering virtual python environment...
Using python3 at /home/francesco/Code/daq/pisuke/daq/venv/bin/python3
Executing: python3 -u daq/daq.py /home/francesco/Code/daq/pisuke/daq/local/system.conf -s
Prepending /home/francesco/Code/daq/pisuke/daq/binhack to PATH
processing arg: /home/francesco/Code/daq/pisuke/daq/local/system.conf
Reading config from /home/francesco/Code/daq/pisuke/daq/local/system.conf
Reading config from misc/system.conf
processing arg: -s
base_conf=misc/module_config.json
default_timeout_sec=350
dhcp_lease_time=500s
ext_addr=192.168.10.10
ext_ctrl=enx645aedf40df6
ext_dpid=0xABCDEF1234
ext_intf=enxb49cdff3015b
ext_loip=192.168.10.@/24
ext_ofip=192.168.10.11/24
ext_ofpt=6653
host_tests=misc/all_tests.conf
initial_dhcp_lease_time=120s
long_dhcp_response_sec=105
monitor_scan_sec=30
run_mode=local
sec_port=24
single_shot=True
site_description="Cisco 9300"
site_path=local/site/
switch_model=CISCO_9300
INFO:daq:pid is 8460
INFO:gcp:No gcp_cred credential specified in config
INFO:config:Loading config from misc/module_config.json
INFO:config:Skipping missing local/site/module_config.json
INFO:topology:No device_specs file specified, skipping...
INFO:runner:Reading test definition file misc/all_tests.conf
INFO:runner:Reading test definition file misc/host_tests.conf
INFO:runner:Reading test definition file subset/pentests/build.conf
INFO:runner:Reading test definition file subset/switch/build.conf
INFO:runner:Reading test definition file subset/connection/build.conf
INFO:runner:Reading test definition file subset/bacnet/build.conf
INFO:runner:Reading test definition file subset/security/build.conf
INFO:runner:Reading test definition file subset/cloud/build.conf
INFO:runner:DAQ RUN id: 6896057c-84da-49ad-8db9-d66eee084159
INFO:runner:Configured with tests pass, fail, ping, bacnet, mudgee, nmap, discover, switch, macoui, bacext, tls, passwords, udmi
INFO:runner:DAQ version 1.1.0-26-g40abe0d
INFO:runner:LSB release Distributor ID: Ubuntu Description: Ubuntu 18.04.4 LTS Release: 18.04 Codename: bionic
INFO:runner:system uname Linux arup-daq-lab 5.3.0-46-generic #38~18.04.1-Ubuntu SMP Tue Mar 31 04:17:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
INFO:network:Activating faucet topology...
INFO:topology:Starting faucet...
INFO:network:Initializing faucitizer...
INFO:network:Configuring external secondary with dpid 737894404660 on intf enxb49cdff3015b
INFO:network:Starting mininet...
INFO:mininet:*** Configuring hosts
INFO:mininet:*** Starting controller
INFO:mininet:controller
INFO:mininet:*** Starting 1 switches
INFO:mininet:pri
INFO:mininet:...
INFO:network:Attaching switch interface ctrl-pri on port 1000
INFO:runner:Waiting for system to settle...
INFO:runner:Entering main event loop.
INFO:runner:See docs/troubleshooting.md if this blocks for more than a few minutes.
INFO:runner:System port 24 on dpid 737894404660 is active True
INFO:runner:Port 5 dpid 737894404660 is now active True
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:Gateway for device group 0040ae10387b not found, initializing base 5...
INFO:gateway:Initializing gateway 0040ae10387b as gw05/50
INFO:gateway:Added networking host gw05 on port 50 at 10.20.0.1
INFO:gateway:Added dummy target dummy05 on port 51 at 10.20.0.2
INFO:gateway:Gateway 5 startup capture gw05-eth0 in container's /tmp/gateway.pcap
INFO:dhcp:DHCP monitor gw05 waiting for replies...
INFO:runner:Test ping gw05->dummy05
INFO:runner:Test ping gw05->dummy05
INFO:runner:Test ping dummy05->gw05
INFO:runner:Test ping dummy05->192.168.84.5
INFO:runner:Test ping gw05->dummy05 from 192.168.84.5
INFO:gateway:Attaching target 5 to gateway group 0040ae10387b
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:host:Configuring device 00:40:ae:10:38:7b as type deltacontrols_o3-din-cpu
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/device_types/deltacontrols_o3-din-cpu/module_config.json
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:host:Device config reloaded: True 00:40:ae:10:38:7b
INFO:config:Writing config to /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/aux/module_config.json
INFO:host:Host 5 running with enabled tests ['pass', 'fail', 'ping', 'switch']
INFO:report:Creating report as inst/reports/report_0040ae10387b_2020-04-13T150224+0000.md
INFO:report:Creating report as inst/reports/report_0040ae10387b_2020-04-13T150224+0000.pdf
INFO:runner:Target port 5 registered 00:40:ae:10:38:7b
INFO:host:Target port 5 initializing...
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:network:Creating mirror pair mirror-5 <-> mirror-5-ext at 1005
INFO:host:Target port 5 startup pcap capture
INFO:host:Target port 5 scan intf mirror-5 for None filter  output in /home/francesco/Code/daq/pisuke/daq/inst/run-port-05/scans/startup.pcap
INFO:network:Directing traffic for 00:40:ae:10:38:7b on port 5: True
INFO:topology:Waiting 5s for network to settle
INFO:host:Target port 5 waiting for ip as 00:40:ae:10:38:7b
INFO:host:Target port 5 using normal DHCP mode
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:IP notify 00:40:ae:10:38:7b is 10.20.15.186 on gw05 (done/70)
INFO:gateway:Ready target 00:40:ae:10:38:7b from gateway group 0040ae10387b
INFO:runner:IP activating target 00:40:ae:10:38:7b
INFO:host:Target port 5 triggered as 10.20.15.186
INFO:runner:Test ping gw05->10.20.15.186
INFO:runner:Test ping gw05->10.20.15.186
INFO:runner:Test ping gw05->10.20.15.186 from 192.168.84.5
INFO:host:Target port 5 monitor scan complete
INFO:gcp:Ignoring /home/francesco/Code/daq/pisuke/daq/inst/run-port-05/scans/startup.pcap upload: not configured
INFO:host:Target port 5 done with base.
INFO:host:Target port 5 background scan for 30s
INFO:host:Target port 5 scan intf mirror-5 for 30 filter  output in /home/francesco/Code/daq/pisuke/daq/inst/run-port-05/scans/monitor.pcap
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:host:Target port 5 scan complete
INFO:host:Target port 5 monitor scan complete
INFO:gcp:Ignoring /home/francesco/Code/daq/pisuke/daq/inst/run-port-05/scans/startup.pcap upload: not configured
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:host:Configuring device 00:40:ae:10:38:7b as type deltacontrols_o3-din-cpu
INFO:config:Writing config to inst/run-port-05/nodes/pass05/tmp/module_config.json
INFO:docker:Target port 5 test pass running
INFO:docker:Target port 5 docker finalize
INFO:docker:Target port 5 test pass passed 1.638414s
INFO:host:Host callback pass/pass05 was 0 with None
INFO:gcp:Ignoring inst/run-port-05/nodes/pass05/activate.log upload: not configured
INFO:gcp:Ignoring inst/run-port-05/nodes/pass05/tmp/module_config.json upload: not configured
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:host:Configuring device 00:40:ae:10:38:7b as type deltacontrols_o3-din-cpu
INFO:config:Writing config to inst/run-port-05/nodes/fail05/tmp/module_config.json
INFO:docker:Target port 5 test fail running
INFO:docker:Target port 5 docker finalize
INFO:docker:Target port 5 test fail failed 1.369355s: 1 None
INFO:host:Host callback fail/fail05 was 1 with None
INFO:gcp:Ignoring inst/run-port-05/nodes/fail05/activate.log upload: not configured
INFO:gcp:Ignoring inst/run-port-05/nodes/fail05/tmp/module_config.json upload: not configured
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:host:Configuring device 00:40:ae:10:38:7b as type deltacontrols_o3-din-cpu
INFO:config:Writing config to inst/run-port-05/nodes/ping05/tmp/module_config.json
INFO:docker:Target port 5 test ping running
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:docker:Target port 5 docker finalize
INFO:docker:Target port 5 test ping passed 20.937261s
INFO:host:Host callback ping/ping05 was 0 with None
INFO:gcp:Ignoring inst/run-port-05/nodes/ping05/tmp/report.txt upload: not configured
INFO:gcp:Ignoring inst/run-port-05/nodes/ping05/activate.log upload: not configured
INFO:gcp:Ignoring inst/run-port-05/nodes/ping05/tmp/module_config.json upload: not configured
INFO:config:Loading config from /home/francesco/Code/daq/pisuke/daq/local/site/mac_addrs/0040ae10387b/module_config.json
INFO:host:Configuring device 00:40:ae:10:38:7b as type deltacontrols_o3-din-cpu
INFO:config:Writing config to inst/run-port-05/nodes/switch05/tmp/module_config.json
INFO:docker:Target port 5 test switch running
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:IP notify 00:40:ae:10:38:7b is 10.20.15.186 on gw05 (done/62)
INFO:runner:DHCP activation group 0040ae10387b already activated
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:IP notify 00:40:ae:10:38:7b is 10.20.15.186 on gw05 (long/255)
INFO:runner:DHCP activation group 0040ae10387b already activated
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
INFO:runner:Port 5 dpid 737894404660 learned 00:40:ae:10:38:7b
ERROR:host:Monitoring timeout for switch after 350s
INFO:docker:Target port 5 test switch terminating
INFO:docker:Target port 5 docker finalize
INFO:host:Host callback switch/unknown was None with Timeout expired
INFO:host:Target port 5 no more tests remaining
INFO:report:Finalizing report report_0040ae10387b_2020-04-13T150224+0000.md
INFO:report:Skipping missing report header template local/site/report_template.md
INFO:report:Copying test report inst/run-port-05/nodes/ping05/tmp/report.txt
INFO:report:Generating HTML for writing pdf report...
INFO:report:Metamorphosising HTML to PDF...
WARNING:weasyprint:Expected a media type, got only/**/screen/**/and/**/(min-width: 480px)
WARNING:weasyprint:Invalid media type " only screen and (min-width: 480px) " the whole @media rule was ignored at 231:1.
WARNING:weasyprint:Expected a media type, got only/**/screen/**/and/**/(min-width: 768px)
WARNING:weasyprint:Invalid media type " only screen and (min-width: 768px) " the whole @media rule was ignored at 236:1.
INFO:report:Copying report to local/site/mac_addrs/0040ae10387b/device_report.md
INFO:report:Also copying report to local/site/mac_addrs/0040ae10387b/device_report.pdf
INFO:gcp:Ignoring inst/reports/report_0040ae10387b_2020-04-13T150224+0000.md upload: not configured
INFO:gcp:Ignoring inst/reports/report_0040ae10387b_2020-04-13T150224+0000.md.json upload: not configured
INFO:runner:Target port 5 finalize: ['05:switch:TimeoutError'] (target set not active)
INFO:runner:Target port 5 cancel 00:40:ae:10:38:7b (#1/0).
INFO:network:Directing traffic for 00:40:ae:10:38:7b on port 5: False
INFO:topology:Waiting 5s for network to settle
INFO:host:Target port 5 terminate, running unknown, trigger False: _target_set_cancel
INFO:network:Deleting mirror pair mirror-5 <-> mirror-5-ext
INFO:gcp:Ignoring /home/francesco/Code/daq/pisuke/daq/inst/run-port-05/scans/ip_triggers.txt upload: not configured
INFO:gateway:Detach target 5 from gateway group 0040ae10387b
INFO:runner:Retiring target gateway 5, 00:40:ae:10:38:7b, 0040ae10387b, 5
INFO:gateway:Terminating gateway 5/0040ae10387b
WARNING:runner:Suppressing future tests because test done in single shot.
INFO:runner:Remaining target sets: []
INFO:stream:Monitoring 0 fds 
WARNING:runner:No active ports remaining (0 monitors), ending test run.
INFO:runner:Stopping network...
INFO:mininet:*** Stopping 1 controllers
INFO:mininet:controller
INFO:mininet:*** Stopping 0 links
INFO:mininet:*** Stopping 1 switches
INFO:mininet:pri
INFO:mininet:*** Stopping 1 hosts
INFO:mininet:gw05
INFO:mininet:*** Done
INFO:runner:Done with runner.
ERROR:runner:Test failures: ['05:switch:TimeoutError']
INFO:daq:DAQ runner returned 1
Cleanup ip link del ctrl-pri
Cleanup ip link del ctrl-swx
Cleanup ovs-vsctl --if-exists del-br ctrl-br
Done with run, exit 1
```

Can you please look into this and give me a hint at where to look in the Java code to try and make them work?

As a general note, it would be useful if in the summary the "pass" result could give evidence to confirm why the test passes (for instance the relevant line in the test result).

Another important thing to add to the test is a readme.md in the test folder (similar for instance to [this one](https://github.com/faucetsdn/daq/blob/master/subset/pentests/readme.md) for the discovery module, where the switch tests are described. It should also include important information to configure the test, for instance the passwords used to set up the switch (manager/friend and admin/password). Can you please do it? Thanks!

**Comment by grafnu** [comment]
On a developer's note, it would be good, I think, to figure out what was
necessary to enable this testing using the bin/test_module capability, so
that it didn't need to run through DAQ proper all the time.  I'm expecting
some "problems" with it because the switch setup requires additional
configuration.  However, I'd be willing to work with somebody on resolving
that... but I'm not in a good position to try it out myself. So, if
somebody wanted to send along the result of "bin/test_module switch" I'd
work with that -- but, I'd recommend doing it in another issue/thread so as
not to pollute this one.  I'll kick up another email thread about it now.

**Comment by jhughesbiot** [comment]
From what I'm seeing from the cisco output and verifying locally with my switch, is that the user you login with via telnet, is immediately enabled.  Once a user logs in, the console ending is expected to be '>' but an enabled user ends with '#'.  The current process for both switches is to login, then enable the user so bypassing the enable step is causing it to sit and wait forever.  You can see that when you login, the console ends with #.  

You should be able to confirm this by logging in directly with a telnet connection.  Once you login, run the command "config t".  If it works, you were enabled during login and if it throws and error like:

> 
> config t
>            ^
> % Invalid input detected at '^' marker.

Then you were logged in and not enabled.

**Comment by jhughesbiot** [comment]
The only thing I saw that was notably different in the configs was that I had

ext_loip=192.168.100.%d/24 

where you had 

ext_loip=192.168.10.@/24

Not really sure if these are equivalent but they both seem to work.

**Comment by grafnu** [comment]
The @ is the "Old" way of doing things, and literally in the code it says
"replace @ with %d" -- at some point I should make the @ an error... but
didn't want to unnecessarily change things on people.  All the
documentation and examples should have %d in them!

**Comment by pisuke** [comment]
> From what I'm seeing from the cisco output and verifying locally with my switch, is that the user you login with via telnet, is immediately enabled. 

@jhughesbiot Indeed, the admin user on the cisco switch is enabled.

So what should the configuration on the switch be?
Can you check the [setup instruction I've written on the wiki](https://github.com/faucetsdn/daq/wiki/Cisco-9300) are correct and add what is missing?

And if the user is not enabled, should the enable password be the same as the ssh password (in this case, well, "password")?

This is another reason to add a `readme.md` file in the test folder to list the setup assumptions.

The behaviour on the AT switch used to be different then, as it was working with the same configuration before. 

Thanks!

**Comment by jhughesbiot** [comment]
The question in the Cisco configuration

> Enter virtual terminal password

By inputting your password there, you do not get prompted for login information over telnet, this should be left blank.  There could be an update (for cisco at least) that checks directly if the user is enabled or just logged in but that would be the quickest way to get testing done.  Right now the code in both switches expects the same password for login and enable so there should only be one password.

As far as the allied switch, nothing is different then how that worked in the past.  The only thing that requires it to change from a configuration aspect is just making sure the switch_model is set right in system.conf and I haven't looked into the output of that switch as I am not familiar with it.  I only maintained it's functionality as is when adding cisco support.

**Comment by pisuke** [comment]
OK, I've changed the wiki page like this:

```
Enter enable secret: cisco

Enter enable password: password

Enter virtual terminal password: 
```

I'll reconfigure my switch and see if I can make it work.

I swear the AT was working with the same config, so something is different.
I'll dig deeper, but concentrating on making the Cisco one work now.

**Comment by pisuke** [comment]
Uhm ... I've removed the password from the switch but the following happens when I attempt to telnet:

```
$ telnet 192.168.10.10
Trying 192.168.10.10...
Connected to 192.168.10.10.
Escape character is '^]'.


Password required, but none set
Connection closed by foreign host.
```
Anything else I need to change?

**Comment by jhughesbiot** [comment]
You still need to set the enable password so if you didn't set that, then you'll need to.  I didn't actually setup this part of the switch i'm working on, this was done by my IT group, I mostly dealt with the openflow side of things once it was turned over.  It might be the enable password needs to be unset and the virtual terminal needs to be set.  Not entirely sure on that but if the result of logging in  via telnet is the user is automatically enabled, neither switch is expecting that state so they don't progress in their testing and are timing out waiting for the enable command.

**Comment by pisuke** [comment]
@jhughesbiot  OK, I've spent the morning configuring the cisco switch. The following config made telnet access work as required by the Cisco switch java code.

```
no enable secret
no enable password
enable secret password
aaa new-model
aaa authentication login default group tacacs+ local line enable
aaa authentication enable default group tacacs+ enable
aaa authorization exec default group tacacs+ if-authenticated
aaa accounting exec default start-stop group tacacs+
```

All working OK with the AT switch again. The `%d` in `system.conf` did the trick.
[device_report_at_x230.pdf](https://github.com/faucetsdn/daq/files/4503920/device_report_at_x230.pdf)
The report output is still missing useful information in the summary when the switch tests pass.
@jhughesbiot can you please add it?

With the Cisco switch however I get an error and an incomplete report (missing the report test output)
[device_report_cisco_9300.pdf](https://github.com/faucetsdn/daq/files/4503932/device_report_cisco_9300.pdf)

Here is the activate.log output from the test with the cisco switch:
[activate.log](https://github.com/faucetsdn/daq/files/4503962/activate.log)

There seems to be a problem with parsing the switch output after the PoE command:
```
java.lang.NumberFormatException: For input string: "4     30.0"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at switchtest.cisco.Cisco9300.validatePowerTests(Cisco9300.java:282)
	at switchtest.cisco.Cisco9300.generateTestResults(Cisco9300.java:72)
	at switchtest.cisco.Cisco9300.parseData(Cisco9300.java:216)
	at switchtest.cisco.Cisco9300.receiveData(Cisco9300.java:178)
	at switchtest.cisco.CiscoSwitchTelnetClientSocket.gatherData(CiscoSwitchTelnetClientSocket.java:53)
	at switchtest.SwitchTelnetClientSocket.lambda$run$1(SwitchTelnetClientSocket.java:95)
	at java.lang.Thread.run(Thread.java:748)
```
@jhughesbiot can you please fix it?

**Comment by jhughesbiot** [comment]
I believe this is happening because there's a space in the device name and the code currently isn't expecting spaces within each column of response data.  Renaming it to something without a space should remove this error.  I'll have to work on a fix for this scenario.

**Comment by pisuke** [comment]
@jhughesbiot it's a bug in my opinion:

```
arup-daq-lab#show interface gigabitethernet1/0/8
GigabitEthernet1/0/8 is up, line protocol is up (connected) 
  Hardware is Gigabit Ethernet, address is 4ce1.7671.d408 (bia 4ce1.7671.d408)
  MTU 1500 bytes, BW 100000 Kbit/sec, DLY 100 usec, 
     reliability 255/255, txload 1/255, rxload 1/255
  Encapsulation ARPA, loopback not set
  Keepalive not set
  Full-duplex, 100Mb/s, media type is 10/100/1000BaseTX
  input flow-control is on, output flow-control is unsupported 
  ARP type: ARPA, ARP Timeout 04:00:00
  Last input never, output 00:00:08, output hang never
  Last clearing of "show interface" counters never
  Input queue: 0/2000/0/0 (size/max/drops/flushes); Total output drops: 0
  Queueing strategy: fifo
  Output queue: 0/40 (size/max)
  5 minute input rate 0 bits/sec, 0 packets/sec
  5 minute output rate 0 bits/sec, 0 packets/sec
     578 packets input, 145924 bytes, 0 no buffer
     Received 523 broadcasts (89 multicasts)
     0 runts, 0 giants, 0 throttles 
     0 input errors, 0 CRC, 0 frame, 0 overrun, 0 ignored
     0 watchdog, 89 multicast, 0 pause input
     0 input packets with dribble condition detected
     2639 packets output, 445545 bytes, 0 underruns
     0 output errors, 0 collisions, 2 interface resets
     0 unknown protocol drops
     0 babbles, 0 late collision, 0 deferred
     0 lost carrier, 0 no carrier, 0 pause output
     0 output buffer failures, 0 output buffers swapped out
arup-daq-lab#show power inline gigabitethernet1/0/8
Interface Admin  Oper       Power   Device              Class Max
                            (Watts)                            
--------- ------ ---------- ------- ------------------- ----- ----
Gi1/0/8   auto   on         30.0    Ieee PD             4     30.0
```
I have no way to name this device in the way it is read by the switch. The hostname is different.
The power indicated is also not the current power but the maximum device power.

I suggest we use the following command instead:
```
arup-daq-lab#show power inline gigabitethernet1/0/8 detail
 Interface: Gi1/0/8
 Inline Power Mode: auto
 Operational status: on
 Device Detected: yes
 Device Type: Ieee PD
 IEEE Class: 4
 Discovery mechanism used/configured: Ieee and Cisco
 Police: off

 Power Allocated 
 Admin Value: 30.0
 Power drawn from the source: 30.0
 Power available to the device: 30.0
 
 Actual consumption
 Measured at the port: 3.2
 Maximum Power drawn by the device since powered on: 3.5
 
 Absent Counter: 0
 Over Current Counter: 0
 Short Current Counter: 0
 Invalid Signature Counter: 0
 Power Denied Counter: 0

 Power Negotiation Used: None
 LLDP Power Negotiation --Sent to PD--      --Rcvd from PD--
   Power Type:          -                    -
   Power Source:        -                    -
   Power Priority:      -                    -
   Requested Power(W):  -                    -
   Allocated Power(W):  -                    -

Four-Pair PoE Supported: No
Spare Pair Power Enabled: No
Four-Pair PD Architecture: N/A
```
This clearly shows that the actual consumption is 3.2 W and this value must be reported in the summary
.

**Comment by jhughesbiot** [comment]
@pisuke 
I agree it's a bug, just explaining what the bug was.  I'll take a look at this alternate command and implement a fix.

**Comment by pisuke** [comment]
Cool, thanks @jhughesbiot for having a look!

**Comment by pisuke** [comment]
Moving the discussion to #375

**Comment by pisuke** [comment]
I have executed a series of tests using the code recently merged in #375 and the output from the module for the `connection.switch.port_link` `connection.switch_port_duplex` and `connection.switch.port_speed` is now aligned with the expectations.

The switch full report output however hasn't been included yet, so I will open another issue for that.

The PoE tests output is correct for a PoE devices but misleading and with wrong test output status for non PoE devices. I will capture this in a separate issue.

For reference, see the minimal reports linked below.

[device_report_non_poe_device.pdf](https://github.com/faucetsdn/daq/files/4649641/device_report_non_poe_device.pdf)

[device_report_poe_device.pdf](https://github.com/faucetsdn/daq/files/4649643/device_report_poe_device.pdf)

Closing this issue.
